File: 2017-minimum-number-of-flips-to-make-the-binary-string-alternating/minimum-number-of-flips-to-make-the-binary-string-alternating.cpp
class Solution {
public:
    int minFlips(string s) {

        string orgStr = s+s;
        string flipStr = s+s;
        int zans = fun(orgStr,flipStr,0);
        int oans = fun(orgStr,flipStr,1);

        return min(zans,oans);
    }

    int fun(string orgStr , string flipStr , int start){
        int curr=start;
        int cnt=0;
        int ans=INT_MAX;

        int s=0,e=0,n=orgStr.size();

        while(e<n){
            int indVa=orgStr[e]-'0';
            if(curr!=indVa){
                cnt++;
                flipStr[e]='0'+curr;
            }
            curr=!curr;
            if(e-s+1 == n/2){
                ans=min(ans,cnt);
                if(orgStr[s]!=flipStr[s]){
                    cnt--;
                }
                s++;
            }
            e++;
        }
        return ans;
    }
};

File: 2017-minimum-number-of-flips-to-make-the-binary-string-alternating/minimum-number-of-flips-to-make-the-binary-string-alternating.java
class Solution {
    public int minFlips(String s) {

        int zans = fun(s, 0);
        int oans = fun(s, 1);

        return Math.min(zans, oans);

    }

    int fun(String str,int start){

        StringBuilder  orgStr = new StringBuilder(str+str);
        StringBuilder flipStr = new StringBuilder(str+str);

        int curr=start;
        int cnt=0;
        int ans=Integer.MAX_VALUE;

        int s=0,e=0,n=orgStr.length();

        while(e<n){
            int indVa=orgStr.charAt(e)-'0';
            if(curr!=indVa){
                cnt++;
                flipStr.setCharAt(e,curr==0?'0':'1');
            }
            curr=curr==0?1:0;
            if(e-s+1 == n/2){
                ans=Math.min(ans,cnt);
                if(orgStr.charAt(s)!=flipStr.charAt(s)){
                    cnt--;
                }
                s++;
            }
            e++;
        }
        return ans;
    }
}